pkg/controller/datavolume-controller.go

```diff
@@ -281,28 +355,107 @@ func (r *DatavolumeReconciler) Reconcile(req reconcile.Request) (reconcile.Resul
 		}
 	}
 
+	cloneStrategy, err := r.getCloneStrategy()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	bindingMode, err := r.getStorageClassBindingMode(datavolume)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	snapshotClassName, err := r.getSnapshotClassForSmartClone(datavolume)
+	doSmartClone := err == nil &&
+		cloneStrategy == cdiv1.CloneStrategySnapshot &&
+		(!isCrossNamespaceClone(datavolume) || *bindingMode == storagev1.VolumeBindingImmediate)
+
 	if !pvcExists {
-		cloneStrategy, err := r.getCloneStrategy()
-		if err != nil {
-			return reconcile.Result{}, err
-		}
-		snapshotClassName, err := r.getSnapshotClassForSmartClone(datavolume)
-		if err == nil && cloneStrategy == cdiv1.CloneStrategySnapshot {
-			r.log.V(3).Info("Smart-Clone via Snapshot is available with Volume Snapshot Class", "snapshotClassName", snapshotClassName)
-			if requeue, err := r.sourceInUse(datavolume); requeue || err != nil {
-				return reconcile.Result{Requeue: requeue}, err
-			}
-			if populated, err := r.isSourcePVCPopulated(datavolume); !populated || err != nil {
-				return reconcile.Result{Requeue: !populated}, err
-			}
-			newSnapshot := newSnapshot(datavolume, snapshotClassName)
-			if err := r.client.Create(context.TODO(), newSnapshot); err != nil {
-				if k8serrors.IsAlreadyExists(err) {
+		if doSmartClone {
+			pvcName := datavolume.Name
+			if isCrossNamespaceClone(datavolume) {
+				pvcName = transferName
+				initialized, err := r.initTransfer(datavolume, pvcName)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+
+				// get get reconciled again v soon
+				if !initialized {
+					return reconcile.Result{},
+						r.updateSmartCloneStatusPhase(cdiv1.CloneScheduled, datavolume, nil)
+				}
+
+				tmpPVC := &corev1.PersistentVolumeClaim{}
+				nn := types.NamespacedName{Namespace: datavolume.Spec.Source.PVC.Namespace, Name: pvcName}
+				if err := r.client.Get(context.TODO(), nn, tmpPVC); err != nil {
+					if !k8serrors.IsNotFound(err) {
+						return reconcile.Result{}, err
+					}
+				} else if tmpPVC.Annotations[AnnCloneOf] == "true" {
+					done, err := r.expand(log, datavolume, tmpPVC)
+					if err != nil {
+						return reconcile.Result{}, err
+					}
+					if !done {
+						// XXX should maybe have expansion phase
+						return reconcile.Result{}, nil
+					}
+
+					// trigger transfer and next reconcile should have pvcExists == true
+					tmpPVC.Annotations[annReadyForTransfer] = "true"
+					tmpPVC.Annotations[AnnPopulatedFor] = datavolume.Name
+					if err := r.client.Update(context.TODO(), tmpPVC); err != nil {
+						return reconcile.Result{}, err
+					}
+
+					// XXX should maybe have transfer phase
```

Comment:
You mean the DV phase should be 'transfer'?

Reply:
Yeah, this should happen pretty quickly so maybe not as important as potential expansion phase though